diff -uNr bitcoin-0.10.2/src/init.cpp bitcoin-0.10.2_bm/src/init.cpp
--- bitcoin-0.10.2/src/init.cpp	2015-05-16 16:37:27.000000000 +0800
+++ bitcoin-0.10.2_bm/src/init.cpp	2015-07-13 18:35:35.000000000 +0800
@@ -287,6 +287,8 @@
     strUsage += "  -proxy=<ip:port>       " + _("Connect through SOCKS5 proxy") + "\n";
     strUsage += "  -seednode=<ip>         " + _("Connect to a node to retrieve peer addresses, and disconnect") + "\n";
     strUsage += "  -timeout=<n>           " + strprintf(_("Specify connection timeout in milliseconds (minimum: 1, default: %d)"), DEFAULT_CONNECT_TIMEOUT) + "\n";
+    strUsage += "  -limitdownloadblocks=<n> " + _("Limit download blocks number from this node's height (default: INT32_MAX)") + "\n";
+    strUsage += "  -outboundconnections=<n> " + _("Max outbound connections, should <= -maxconnections (default: 8)") + "\n";
 #ifdef USE_UPNP
 #if USE_UPNP
     strUsage += "  -upnp                  " + _("Use UPnP to map the listening port (default: 1 when listening)") + "\n";
@@ -625,10 +627,15 @@
             LogPrintf("AppInit2 : parameter interaction: -zapwallettxes=<mode> -> setting -rescan=1\n");
     }

+    // nLimitDownloadBlocks
+    nLimitDownloadBlocks = GetArg("-limitdownloadblocks", 2147483647);
+
     // Make sure enough file descriptors are available
     int nBind = std::max((int)mapArgs.count("-bind") + (int)mapArgs.count("-whitebind"), 1);
     nMaxConnections = GetArg("-maxconnections", 125);
     nMaxConnections = std::max(std::min(nMaxConnections, (int)(FD_SETSIZE - nBind - MIN_CORE_FILEDESCRIPTORS)), 0);
+    nMaxOutboundConnections = GetArg("-outboundconnections", 8);
+    nMaxOutboundConnections = std::min(nMaxConnections, nMaxOutboundConnections);
     int nFD = RaiseFileDescriptorLimit(nMaxConnections + MIN_CORE_FILEDESCRIPTORS);
     if (nFD < MIN_CORE_FILEDESCRIPTORS)
         return InitError(_("Not enough file descriptors available."));
diff -uNr bitcoin-0.10.2/src/main.cpp bitcoin-0.10.2_bm/src/main.cpp
--- bitcoin-0.10.2/src/main.cpp	2015-05-16 16:37:27.000000000 +0800
+++ bitcoin-0.10.2_bm/src/main.cpp	2015-07-13 18:35:35.000000000 +0800
@@ -53,6 +53,7 @@
 bool fIsBareMultisigStd = true;
 bool fCheckBlockIndex = false;
 unsigned int nCoinCacheSize = 5000;
+int32_t nLimitDownloadBlocks = 2147483647;


 /** Fees smaller than this (in satoshi) are considered zero fee (for relaying and mining) */
@@ -3901,6 +3902,14 @@
             pindex = chainActive.Next(pindex);
         int nLimit = 500;
         LogPrint("net", "getblocks %d to %s limit %d from peer=%d\n", (pindex ? pindex->nHeight : -1), hashStop==uint256(0) ? "end" : hashStop.ToString(), nLimit, pfrom->id);
+
+        int nBestHeight = (int)chainActive.Height();
+        if (pindex && nBestHeight > nLimitDownloadBlocks && pindex->nHeight < nBestHeight - nLimitDownloadBlocks) {
+            pfrom->fDisconnect = true;
+            return error("getblocks error : too old block, request height: %d, should more than %d",
+                         pindex->nHeight, nBestHeight - nLimitDownloadBlocks);
+        }
+
         for (; pindex; pindex = chainActive.Next(pindex))
         {
             if (pindex->GetBlockHash() == hashStop)
diff -uNr bitcoin-0.10.2/src/main.h bitcoin-0.10.2_bm/src/main.h
--- bitcoin-0.10.2/src/main.h	2015-05-16 16:37:27.000000000 +0800
+++ bitcoin-0.10.2_bm/src/main.h	2015-07-13 18:35:35.000000000 +0800
@@ -128,6 +128,7 @@
 extern bool fIsBareMultisigStd;
 extern bool fCheckBlockIndex;
 extern unsigned int nCoinCacheSize;
+extern int32_t nLimitDownloadBlocks;
 extern CFeeRate minRelayTxFee;

 /** Best header we've seen so far (used for getheaders queries' starting points). */
diff -uNr bitcoin-0.10.2/src/net.cpp bitcoin-0.10.2_bm/src/net.cpp
--- bitcoin-0.10.2/src/net.cpp	2015-05-16 16:37:27.000000000 +0800
+++ bitcoin-0.10.2_bm/src/net.cpp	2015-07-13 18:35:56.000000000 +0800
@@ -53,8 +53,6 @@
 using namespace std;

 namespace {
-    const int MAX_OUTBOUND_CONNECTIONS = 8;
-
     struct ListenSocket {
         SOCKET socket;
         bool whitelisted;
@@ -78,6 +76,7 @@
 static std::vector<ListenSocket> vhListenSocket;
 CAddrMan addrman;
 int nMaxConnections = 125;
+int nMaxOutboundConnections = 8;
 bool fAddressesInitialized = false;

 vector<CNode*> vNodes;
@@ -871,7 +870,7 @@
                     if (nErr != WSAEWOULDBLOCK)
                         LogPrintf("socket error accept failed: %s\n", NetworkErrorString(nErr));
                 }
-                else if (nInbound >= nMaxConnections - MAX_OUTBOUND_CONNECTIONS)
+                else if (nMaxConnections > nMaxOutboundConnections && nInbound >= nMaxConnections - nMaxOutboundConnections)
                 {
                     CloseSocket(hSocket);
                 }
@@ -1629,7 +1628,7 @@

     if (semOutbound == NULL) {
         // initialize semaphore
-        int nMaxOutbound = min(MAX_OUTBOUND_CONNECTIONS, nMaxConnections);
+        int nMaxOutbound = min(nMaxOutboundConnections, nMaxConnections);
         semOutbound = new CSemaphore(nMaxOutbound);
     }

@@ -1671,7 +1670,7 @@
     LogPrintf("StopNode()\n");
     MapPort(false);
     if (semOutbound)
-        for (int i=0; i<MAX_OUTBOUND_CONNECTIONS; i++)
+        for (int i=0; i<nMaxOutboundConnections; i++)
             semOutbound->post();

     if (fAddressesInitialized)
diff -uNr bitcoin-0.10.2/src/net.h bitcoin-0.10.2_bm/src/net.h
--- bitcoin-0.10.2/src/net.h	2015-05-16 16:37:27.000000000 +0800
+++ bitcoin-0.10.2_bm/src/net.h	2015-07-13 18:35:35.000000000 +0800
@@ -126,6 +126,7 @@
 extern uint64_t nLocalHostNonce;
 extern CAddrMan addrman;
 extern int nMaxConnections;
+extern int nMaxOutboundConnections;

 extern std::vector<CNode*> vNodes;
 extern CCriticalSection cs_vNodes;
